### PR TITLE
Switch from AutomaticFeature to Feature and pass in via --feature

### PR DESCRIPTION
**Pull request by zakkak**
While the Feature interface is API, the annotation AutomaticFeature is
not in the API. That is intentional: if a feature is necessary, then it
should be added via the --features=... option in a
native-image.properties file.

Source: https://github.com/oracle/graal/discussions/4616

Implements 1st part of https://github.com/quarkusio/quarkus/issues/25943

Marking as Draft till tested completes on my fork.

### CODE REVIEW

**Comment by gastaldi** [comment]
@zakkak there are other `Feature` classes implemented using `@AutomaticFeature` in the codebase (extensions like Liquibase, Hibernate, to name a few). 

Should they be migrated to this strategy too?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -642,6 +656,17 @@ public NativeImageInvokerInfo build() {
                     enableHttpsUrlHandler = true;
                 }
 
+                List<String> featuresList = new ArrayList<>(graalVmFeatures == null ? 3 : graalVmFeatures.size() + 3);
+                featuresList.add(NativeImageFeatureStep.GRAAL_FEATURE);
+                featuresList.add(ResourcesFeature.class.getName());
+                featuresList.add(DisableLoggingFeature.class.getName());
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:662]: Can we have them produced elsewhere? IMHO It would be much better if this part could just consume them

Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: Done

**Review by gastaldi — APPROVED** [review]

**Review by gastaldi — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * Represents a GraalVM feature to be passed to native-image through the {@code --features} options.
+ */
+public final class GraalVmFeatureBuildItem extends MultiBuildItem {
+
+    private final String qualifiedName;
+
+    public GraalVmFeatureBuildItem(Class featureClass) {
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java:14]:
```suggestion
    public GraalVmFeatureBuildItem(Class<? extends Feature> featureClass) {
```

**Comment by gastaldi** [comment]
Looks good to me, added a minor suggestion for better type check purposes

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package io.quarkus.deployment.builditem;
+
+import java.util.Objects;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * Represents a GraalVM feature to be passed to native-image through the {@code --features} options.
+ */
+public final class GraalVmFeatureBuildItem extends MultiBuildItem {
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java:10]:
```suggestion
public final class GraalVMFeatureBuildItem extends MultiBuildItem {
```

**Review by gastaldi — CHANGES_REQUESTED** [review]
We use `GraalVM` in other parts, so better follow the pattern :)

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java (hunk 3 above)
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java]: Good catch. I changed all `[Gg]raalVm` prefixes to `[Gg]raalVM`

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 4]
```diff
@@ -565,6 +582,11 @@ public Builder setUnsupportedOSes(
                 return this;
             }
 
+            public Builder setgraalVMFeatures(List<GraalVMFeatureBuildItem> graalVMFeatures) {
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:585]:
```suggestion
            public Builder setGraalVMFeatures(List<GraalVMFeatureBuildItem> graalVMFeatures) {
```

**Comment by sberyozkin** [comment]
@stuartwdouglas Can you also have a look please ?

**Review by gastaldi — APPROVED** [review]
LGTM

File: extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java [hunk 5]
```diff
@@ -251,10 +267,9 @@ public void write(String s, byte[] bytes) {
                 }
             }, "io.quarkus.security.BouncyCastleFeature", null, Object.class.getName(),
                     org.graalvm.nativeimage.hosted.Feature.class.getName());
-            file.addAnnotation("com.oracle.svm.core.annotate.AutomaticFeature");
 
-            MethodCreator beforeAn = file.getMethodCreator("beforeAnalysis", "V",
-                    org.graalvm.nativeimage.hosted.Feature.BeforeAnalysisAccess.class.getName());
+            MethodCreator beforeAn = file.getMethodCreator("afterRegistration", "V",
+                    org.graalvm.nativeimage.hosted.Feature.AfterRegistrationAccess.class.getName());
```
Comment (gastaldi) [inline extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java:272]: Is this change intentional?

Comment (zakkak) [reply extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java:272]: Yes. Apparently without this change the BC provider doesn't get registered on time. I assume this has to do with the ordering that the different Features are being ran which is impacted by removing the `AutomaticFeature` annotation from our features. Furthermore, it should have been that way anyway, see https://github.com/oracle/graal/issues/2800#issuecomment-702480444.

Comment (gastaldi) [reply extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java:272]: Can you also rename the variable to reflect thia change?

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java (hunk 3 above)
Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java]: In general all our native image build items are prefixed with `NativeImage` rather than `GraalVM`.

Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/GraalVmFeatureBuildItem.java]: Fixed

**Review by gsmet — COMMENTED** [review]
It looks good to me except for one small interrogation regarding nested classes and escaping.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 6]
```diff
@@ -642,6 +664,14 @@ public NativeImageInvokerInfo build() {
                     enableHttpsUrlHandler = true;
                 }
 
+                assert nativeImageFeatures != null && !nativeImageFeatures.isEmpty()
+                        : "GraalVM features can't be empty, quarkus core is using some";
+                List<String> featuresList = new ArrayList<>(nativeImageFeatures.size());
+                for (NativeImageFeatureBuildItem nativeImageFeature : nativeImageFeatures) {
+                    featuresList.add(nativeImageFeature.getQualifiedName());
+                }
+                nativeImageArgs.add("--features=" + String.join(",", featuresList));
```
[the remark this answers is not in the document]
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:673]: That's an interesting question, I'll give it a try. 

PS: Why would one define a feature as a nested class though?

File: extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java [hunk 7]
```diff
@@ -231,6 +233,20 @@ void recordBouncyCastleProviders(SecurityProviderRecorder recorder,
...
+
+        Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider = getOne(bouncyCastleJsseProviders);
+        Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider = getOne(bouncyCastleProviders);
+
+        if (bouncyCastleJsseProvider.isPresent() || bouncyCastleProvider.isPresent()) {
+            return new NativeImageFeatureBuildItem("io.quarkus.security.BouncyCastleFeature");
+        }
+        return null;
+    }
+
```
Comment (sberyozkin) [inline extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java:249]: I understand the difference now, thanks

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:673]: > PS: Why would one define a feature as a nested class though?

People are weird :).

[the remark this answers is not in the document]
Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:673]: That's a good question, and possible :)

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 8]
```diff
@@ -642,6 +664,14 @@ public NativeImageInvokerInfo build() {
                     enableHttpsUrlHandler = true;
                 }
 
+                assert nativeImageFeatures != null && !nativeImageFeatures.isEmpty()
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:667]: `assert`?

We don't really use it in the codebase

Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:667]: Should I add an if-statement and through an exception instead?

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:667]: Yes please

**Comment by geoand** [comment]
This needs a rebase

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:673]: According to my tests the `$` is being properly escaped **but** we still can't use nested classes for defining a `Feature` even when using `@AutomaticFeature` (i.e. without this PR).

`native-image` fails with:
```
Error: Error instantiating Feature class io.quarkus.runtime.graal.DisableLoggingFeature$NestedDisableLoggingFeature. Ensure the class is not abstract and has a no-argument constructor.
com.oracle.svm.core.util.UserError$UserException: Error instantiating Feature class io.quarkus.runtime.graal.DisableLoggingFeature$NestedDisableLoggingFeature. Ensure the class is not abstract and has a no-argument constructor.
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:84)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.registerFeature(FeatureHandler.java:180)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.registerFeatures(FeatureHandler.java:128)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:836)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:559)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:519)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
Caused by: java.lang.NoSuchMethodException: io.quarkus.runtime.graal.DisableLoggingFeature$NestedDisableLoggingFeature.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3349)
	at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2553)
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ReflectionUtil.lookupConstructor(ReflectionUtil.java:81)
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ReflectionUtil.lookupConstructor(ReflectionUtil.java:76)
	at org.graalvm.nativeimage.base/com.oracle.svm.util.ReflectionUtil.newInstance(ReflectionUtil.java:95)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.registerFeature(FeatureHandler.java:178)
	... 7 more
```

Explicitly defining the default public constructor doesn't help either.

[the remark this answers is not in the document]
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:673]: This is essentially happening because inner classes don't actually have "a no-argument constructor" since the seemingly no-argument default constructor of inner classes takes as argument an instance of the outer class. However, GraalVM doesn't handle this case so it doesn't support defining nested `Feature`s.

[the remark this answers is not in the document]
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:681]: static nested features should work though, no?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 8 above)
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java]: Done

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:681]: Good point! In that case everything works as expected (`$` gets escaped and the feature gets registered).

E.g.:
```
...
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] /home/zakkak/code/graal/sdk/latest_graalvm_home/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=3 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Duser.language=en -J-Duser.country=IE -J-Dfile.encoding=UTF-8 --features=io.quarkus.runner.Feature,io.quarkus.runtime.graal.ResourcesFeature,io.quarkus.runtime.graal.DisableLoggingFeature\$NestedDisableLoggingFeature -H:-ParseOnce -J--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -J-Djava.awt.headless=true -H:FallbackThreshold=0 --link-at-build-time -H:+ReportExceptionStackTraces -g -H:DebugInfoSourceSearchPath=app-sources -H:-AddAllCharsets -H:EnableURLProtocols=http -H:NativeLinkerOption=-no-pie -H:-UseServiceLoaderFeature -H:+StackTrace -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jni=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.localization=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.base/com.oracle.svm.util=ALL-UNNAMED quarkus-integration-test-awt-999-SNAPSHOT-runner -jar quarkus-integration-test-awt-999-SNAPSHOT-runner.jar
...
[1/7] Initializing...                                                                                   (17.0s @ 0.16GB)
 Version info: 'GraalVM 22.2.0-dev Java 11 CE'
 Java version info: '11.0.16+5-jvmci-22.2-b03'
 C compiler: gcc (redhat, x86_64, 12.1.1)
 Garbage collector: Serial GC
 5 user-specific feature(s)
 - io.quarkus.awt.runtime.graal.AwtFeature
 - io.quarkus.runner.Feature
 - io.quarkus.runtime.graal.DisableLoggingFeature$NestedDisableLoggingFeature
 - io.quarkus.runtime.graal.ResourcesFeature
 - org.graalvm.home.HomeFinderFeature: Finds GraalVM paths and its version number
```

Notice the `--features=io.quarkus.runner.Feature,io.quarkus.runtime.graal.ResourcesFeature,io.quarkus.runtime.graal.DisableLoggingFeature\$NestedDisableLoggingFeature` parameter.

**Review by gsmet — APPROVED** [review]
